cluster-autoscaler/cloudprovider/openshiftmachineapi/machineapi_controller_test.go

```diff
@@ -371,3 +394,533 @@ func TestMachinesInMachineSet(t *testing.T) {
+	for i, tc := range []struct {
+		description string
+		annotations map[string]string
+		count       int
+	}{{
+		description: "errors because minSize is invalid",
+		annotations: map[string]string{
+			nodeGroupMinSizeAnnotationKey: "-1",
+			nodeGroupMaxSizeAnnotationKey: "0",
+		},
+	}, {
+		description: "errors because maxSize is invalid",
+		annotations: map[string]string{
+			nodeGroupMinSizeAnnotationKey: "0",
+			nodeGroupMaxSizeAnnotationKey: "-1",
+		},
+	}, {
+		description: "errors because minSize > maxSize",
+		annotations: map[string]string{
+			nodeGroupMinSizeAnnotationKey: "1",
+			nodeGroupMaxSizeAnnotationKey: "0",
+		},
+	}, {
+		description: "errors because maxSize < minSize",
+		annotations: map[string]string{
+			nodeGroupMinSizeAnnotationKey: "1",
+			nodeGroupMaxSizeAnnotationKey: "0",
+		},
+	}, {
+		annotations: map[string]string{
+			nodeGroupMaxSizeAnnotationKey: "10",
+		},
+		count: 1,
+	}, {
+		annotations: map[string]string{
+			nodeGroupMinSizeAnnotationKey: "1",
+			nodeGroupMaxSizeAnnotationKey: "10",
+		},
+		count: 1,
+	}} {
+		machineSet := &v1beta1.MachineSet{
+			TypeMeta: v1.TypeMeta{
+				Kind: "MachineSet",
+			},
+			ObjectMeta: v1.ObjectMeta{
+				Name:        fmt.Sprintf("machineset-%d", i),
+				Namespace:   "test-namespace",
+				Annotations: tc.annotations,
+			},
+		}
+
+		controller, stop := mustCreateTestController(t, testControllerConfig{
+			machineObjects: []runtime.Object{
+				machineSet,
+			},
+		})
+		defer stop()
+
+		nodegroups, err := controller.nodeGroups()
+		if tc.count == 0 && err == nil {
+			t.Fatalf("expected an error")
+		}
+
+		if l := len(nodegroups); l != tc.count {
+			t.Errorf("expected %v, got %v", tc.count, l)
+		}
+	}
+}
+
+func TestControllerNodeGroupForNodeWithMissingNode(t *testing.T) {
+	machine := &v1beta1.Machine{
+		TypeMeta: v1.TypeMeta{
+			Kind: "Machine",
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name:      "machine",
+			Namespace: "test-namespace",
+			OwnerReferences: []v1.OwnerReference{{
+				Kind: "MachineSet",
+				UID:  "ec21c5fb-a3d5-a45f-887b-6b49aa8fc218",
+				Name: "testMachineSet",
+			}},
+		},
+	}
+
+	machineSet := &v1beta1.MachineSet{
+		TypeMeta: v1.TypeMeta{
+			Kind: "MachineSet",
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name:      "machineset",
+			Namespace: "test-namespace",
+			UID:       "ec21c5fb-a3d5-a45f-887b-6b49aa8fc218",
+		},
+	}
+
+	controller, stop := mustCreateTestController(t, testControllerConfig{
+		machineObjects: []runtime.Object{
+			machine,
+			machineSet,
+		},
+	})
+	defer stop()
+
+	ng, err := controller.nodeGroupForNode(&apiv1.Node{
+		TypeMeta: v1.TypeMeta{
+			Kind: "Node",
+		},
+		ObjectMeta: v1.ObjectMeta{
+			Name: "node",
+		},
+	})
+
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+
+	if ng != nil {
+		t.Fatalf("unexpected nodegroup: %v", ng)
+	}
+}
+
+func TestControllerNodeGroupForNodeWithMissingMachineOwner(t *testing.T) {
+	node := &apiv1.Node{
```

Comment:
Worth moving the definition under a function and calling the function everywhere the node is used instead of duplicating the definition.